src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java

```diff
@@ -102,20 +102,31 @@ public static void finalizeRegion(final AssemblyRegion region,
-                // TODO unclipping soft clips may introduce bases that aren't in the extended region if the unclipped bases
-                // TODO include a deletion w.r.t. the reference.  We must remove kmers that occur before the reference haplotype start
-                .map(read -> dontUseSoftClippedBases || ! ReadUtils.hasWellDefinedFragmentSize(read) ?
-                    ReadClipper.hardClipSoftClippedBases(read) : ReadClipper.revertSoftClippedBases(read))
-                .map(read -> softClipLowQualityEnds ? ReadClipper.softClipLowQualEnds(read, minTailQualityToUse) :
-                        ReadClipper.hardClipLowQualEnds(read, minTailQualityToUse))
-                .filter(read -> read.getStart() <= read.getEnd())
-                .map(read -> read.isUnmapped() ? read : ReadClipper.hardClipAdaptorSequence(read))
-                .filter(read ->  !read.isEmpty() && read.getCigar().getReadLength() > 0)
-                .map(read -> ReadClipper.hardClipToRegion(read, region.getPaddedSpan().getStart(), region.getPaddedSpan().getEnd() ))
-                .filter(read -> read.getStart() <= read.getEnd() && read.getLength() > 0 && read.overlaps(region.getPaddedSpan()))
-                .sorted(new ReadCoordinateComparator(readsHeader)) // TODO: sort may be unnecessary here
-                .collect(Collectors.toList());
+
+        final List<GATKRead> readsToUse = new ArrayList<>();
+        for (GATKRead originalRead : region.getReads()) {
+            // TODO unclipping soft clips may introduce bases that aren't in the extended region if the unclipped bases
+            // TODO include a deletion w.r.t. the reference.  We must remove kmers that occur before the reference haplotype start
+            GATKRead read = (dontUseSoftClippedBases || ! ReadUtils.hasWellDefinedFragmentSize(originalRead) ?
+                    ReadClipper.hardClipSoftClippedBases(originalRead) : ReadClipper.revertSoftClippedBases(originalRead));
+            read = (softClipLowQualityEnds ? ReadClipper.softClipLowQualEnds(read, minTailQualityToUse) :
+                    ReadClipper.hardClipLowQualEnds(read, minTailQualityToUse));
+
+            if (read.getStart() <= read.getEnd()) {
+                read = (read.isUnmapped() ? read : ReadClipper.hardClipAdaptorSequence(read));
+
+                if (!read.isEmpty() && read.getCigar().getReadLength() > 0) {
+                    read = ReadClipper.hardClipToRegion(read, region.getPaddedSpan().getStart(), region.getPaddedSpan().getEnd() );
+
+                    if (read.getStart() <= read.getEnd() && read.getLength() > 0 && read.overlaps(region.getPaddedSpan())) {
+                        // NOTE: here we make a defensive copy of the read if it has not been modified by the above operations
+                        // which might only make copies in the case that the read is actually clipped
+                        readsToUse.add(read == originalRead? read.copy() : read);
```

Comment:
The `copy()` method performs a shallow copy -- it ends up calling `SAMRecord.clone()`, which copies all fields in the SAMRecord as if by assignment (the exception is the attributes array, which is explicitly copied). So the `byte[]` arrays for the bases and quals point to the same memory location in the copy as in the original. Is this a problem? Do we later modify the bases/quals in-place somewhere, or do we always copy the bases/quals arrays due to the defensive copies in `SAMRecordToGATKReadAdapter`?

Reply:
We do modify them later, specifically when we call `cleanOverlappingReadPairs()` we modify the base qualities in place for reads that overlap and if any of those reads have not been clipped bby the clipping code here this could result in the next assembly region having invalid/wrong base qualities for the same read. Given how that method is structured it is non-trivial to refactor it to make the copy only in the event it needs to be modified and it seemed easier to just put a check in to make sure that every read is deeply copied at least once.

Reply:
@jamesemery I disagree with your assessment that `cleanOverlappingReadPairs()` changes the bases/quals in-place -- it calls `getBases()` and `getBaseQualities()` on the read, which perform defensive copies. The question is: is there any code that calls `getBasesNoCopy()` and/or `getBaseQualitiesNoCopy()` and truly modifies the bases/quals arrays in place?

Reply:
I see that the `PileupReadErrorCorrector` calls `getBaseQualitiesNoCopy()` and `getBasesNoCopy()` and then modifies the bases/quals in-place. This could have the effect of modifying the base/qual arrays in the original reads that will be used in subsequent assembly regions. I think as part of this PR we should patch `PileupReadErrorCorrector` to call the getters that make copies, and then call the setter methods to update the bases/quals.